### PR TITLE
fix(webview/slack): scanner spawn + notifications + post-pushState target match (#1016)

### LINKED ISSUE #1016: [Feature] webview: Slack — full end-to-end parity with native app
## Summary

Ensure the Slack webview (`https://app.slack.com/client/`) works identically to the native Slack desktop app: login, messaging, channels, threads, reactions, Huddles (voice/video), file sharing, native OS notifications, and continuous memory ingestion.

## Problem

While Slack is a CDP-migrated provider (Network MITM + IDB + DOM snapshot scanner), there are gaps versus the native app experience: Huddle popup whitelisting needs verification under all edge cases, and end-to-end coverage of all Slack features has not been formally validated.

## Solution

Validate and fix every Slack feature path. The scanner (`slack_scanner/`) uses CDP exclusively (zero JS injection). IDB scan interval is 30s. Messages are grouped by (channel_id, day) and upserted via `openhuman.memory_doc_ingest`.

## Acceptance criteria

- [ ] **Auth** — Google SSO, email/password, SAML SSO all complete without manual intervention; session persists across app restarts
- [ ] **Messaging** — Send and receive messages in channels, DMs, and group DMs; threads render and are navigable
- [ ] **Reactions & emoji** — Emoji picker opens, reactions post correctly
- [ ] **File sharing** — Upload and download files; image previews render inline
- [ ] **Huddles** — Huddle popup spawns correctly (`about:blank` → huddle URL whitelisted); audio/video functional; Huddle end cleans up popup
- [ ] **Notifications** — Native OS notifications fire for new messages; respect per-channel mute settings; DND honored; `notification_settings` toggle works
- [ ] **Memory ingestion** — IDB scanner extracts messages, users, channels, workspace name; `memory_doc_ingest` posts land in core; messages grouped by (channel_id, day)
- [ ] **DOM snapshot** — `dom_snapshot.rs` fast-tick captures unread badges and channel list accurately
- [ ] **Multi-workspace** — Switching workspaces within the same webview works; scanner tracks correct team_id
- [ ] **Session persistence** — Tab switch away and back preserves warm session; no re-auth required
- [ ] **Search** — Slack's built-in search functional within the webview
- [ ] **Navigation** — External links open in system browser; internal links stay in webview; allowed hosts (`slack.com`, `slack-edge.com`, `slack-imgs.com`, `slack-files.com`) all resolve

## Known issues / limitations

- Huddle popup handling uses `about:blank` whitelisting which may break if Slack changes their popup flow
- Scanner can be force-disabled via `OPENHUMAN_DISABLE_SLACK_SCANNER=1` env var (dev escape hatch)
- IDB scan interval is 30s — messages within that window may be delayed vs native push

## Related

- Scanner: `app/src-tauri/src/slack_scanner/` (mod.rs, idb.rs, extract.rs, dom_snapshot.rs)
- Huddle popup logic: `app/src-tauri/src/webview_accounts/mod.rs`
- Manifest: `app/src-tauri/recipes/slack/manifest.json`

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

- Slack memory ingest was silently dead on `main` because `lib.rs:998` double-Arc-wrapped `slack_scanner::ScannerRegistry::new()` (which already returns `Arc<Self>`); the registry lookup never matched and the scanner never spawned.
- Native Slack notifications were silent because the existing JS shim only masked `Notification.permission` for the page check — CEF's real Notification path silently no-op'd without an actual permission grant.
- Once the scanner started running, a second issue surfaced: Slack's client router does `pushState('/client/<workspace>/<channel>')` shortly after first load and strips the `#openhuman-account-<id>` fragment our matcher relied on, so every scan tick after pushState failed.
- All three are fixed in this PR (one micro-commit each), plus an audit-deliverable doc capturing the smoke-walk against `pnpm dev:app` on the merged branch tip.
- Three pre-audit hypotheses (allowed_hosts mismatch, default-notifications-on, per-day grouping) were disproved or made moot during smoke and intentionally dropped from scope.

## Problem

Issue #1016 asks for end-to-end Slack parity. Static audit flagged five candidate gaps; manual smoke against `pnpm dev:app` on the branch tip exposed three real bugs and disproved or deferred the rest.

1. **Memory ingestion completely dead (#7)** — `[webview-accounts] slack ScannerRegistry not in app state` warned every time a Slack account opened. RPC `openhuman.memory_recall_memories {namespace:"slack-web:<acct>"}` returned an empty array. The scanner registration in `lib.rs:998` did `manage(std::sync::Arc::new(slack_scanner::ScannerRegistry::new()))`, but `ScannerRegistry::new()` already returns `Arc<Self>` (`mod.rs:744-746`), so the managed type was `Arc<Arc<…>>` while every lookup site uses `try_state::<Arc<…>>()`. Whatsapp/discord/telegram on the surrounding lines (997/999/1000) were already correct; only slack regressed.
2. **Notifications never reach macOS (#6)** — `[cef-helper-notify] installed shims browser_id=… origin=https://app.slack.com/client` confirms the shim ran, but no `forward_native_notification` events fired when DMs arrived. The shim makes `Notification.permission` look granted to page scripts (so providers stop showing "enable notifications" banners), but at the C++ layer the real CEF Notification path silently no-ops when no actual permission has been granted, so `new Notification(...)` never reached our notify-IPC hook.
3. **Scanner can't find the CDP target after Slack's pushState** — surfaced once Bug A was fixed and the scanner actually started running. The matcher `t.url.starts_with(prefix) && t.url.ends_with(fragment)` worked on the first tick (right after `Page.navigate` to the fragment URL) and then failed every subsequent tick because Slack's client router pushed `/client/<workspace>/<channel>` over the URL. Without this, the Bug A fix would have looked like a no-op end-to-end.

## Solution

Three micro-commits, ordered trivial → bounded:

1. **`fix(webview/slack): unwrap double-Arc on slack ScannerRegistry (#1016)`** — drop the redundant outer `std::sync::Arc::new(...)` wrapper at `lib.rs:998` so the managed-state type matches the lookup. Mirrors the working pattern already used for whatsapp/discord/telegram on the surrounding lines.
2. **`fix(webview/cdp): grant notifications via Browser.grantPermissions per provider origin (#1016)`** — issue a browser-scoped `Browser.grantPermissions(["notifications"])` CDP call (`sessionId=None`) right after the page session attaches, scoped to the provider's origin via a new `origin_of()` helper that pulls `scheme://host[:port]` out of `real_url`. Applies to every CDP-attached migrated provider (slack, whatsapp, telegram, discord, browserscan); slack is the regression that surfaced the gap, the others get it for free.
3. **`fix(webview/slack): match same-origin page target after pushState strips fragment (#1016)`** — relax the matcher in both `scan_once` and `dom_scan_once`: try the strict (prefix + fragment) match first, fall back to any same-origin Slack page target. Per-account `data_directory` isolation guarantees one Slack page-target per origin per account in the helper's target list, so the broader match doesn't risk cross-account misattribution even with multiple Slack accounts open.

The doc commit (`docs(qa): add Slack webview parity audit matrix (#1016)`) captures the smoke procedure, the per-criterion verdicts (9 pass / 3 fail → 9 pass after these three fixes / 1 skipped), and the deferred items.

### Out of scope (file as separate issues if not already tracked)

- **Huddle popup stays blank** (#1016 criterion 5) — `popup = window.open('about:blank'); popup.location = huddleURL` cross-window write isn't honored by CEF child popups. Fix shape: CDP listener on parent main-frame nav matching `app.slack.com/huddle/`, force-navigate child via `Page.navigate`. Unbounded debug, deferred.
- **Parent CEF webview goes blank after huddle popup interaction + tab switch** — likely shares root cause with the huddle popup; verify after that lands.
- **`pnpm dev:app` `restart_app` regression on PR #1007** — orthogonal, affects all dev:app sessions (not just Slack). Identity-flip handler fires unconditionally in dev mode, vite dies with `cargo tauri dev` parent, exit 143. Worked around via skip-worktree dev hack in this branch (not committed).
- **In-app Slack logout removes the sidebar tile but cookies persist** — UX inconsistency; re-add lands already signed in.

### Pre-audit hypotheses dropped after smoke

- `allowed_hosts: ["slack.com", "slack-edge.com", "slackb.com"]` is correct; the issue body's `slack-imgs.com` / `slack-files.com` claim was stale. File sharing and image previews work end-to-end.
- Reactions, threads, per-channel mute, per-day memory grouping, and the `notification_settings = true` default are all gated on the scanner actually running and were either auto-resolved or are not actually broken once Bug A and Bug D land. None are touched in this PR.

## Submission Checklist

- [x] **Unit tests** — `cargo test --lib slack_scanner` is green (3 passed). The scanner-side fixes are CDP plumbing changes verified by manual smoke + log evidence; no new unit tests added because the matchers are purely integration-shaped.
- [x] **E2E / integration** — Manual smoke against `pnpm dev:app` on this branch tip (macOS arm64) walked all 12 acceptance criteria from #1016 and verified the three fixes end-to-end. Memory recall via `openhuman.memory_recall_memories` returned 7 docs (channels + DMs) post-fix; macOS toast confirmed when OpenHuman was unfocused. Procedure documented in `docs/qa/SLACK-PARITY.md`.
- [x] **Doc comments** — comments added explaining the `indexName`-style trap (`cdp/session.rs` Browser.grantPermissions block), the pushState fragment-survival assumption (`slack_scanner/mod.rs` matcher fallback), and the `origin_of()` helper invariants.
- [x] **Inline comments** — root-cause + non-obvious-trade-off comments added to each fix site; commit messages cite line numbers where the trap lives.

## Impact

- **Runtime**: macOS / Windows / Linux desktop — affects every CEF-attached migrated provider for the notification grant path; affects only Slack for the scanner-spawn and target-match changes.
- **Security**: the notification grant is per-provider-origin (not blanket); we already control the origin via the per-account CDP session. No new attack surface.
- **Compatibility**: no migration. Existing accounts that were silently logging zero memory will start ingesting on first scan tick after upgrade — the memory_doc_ingest path uses upserts keyed by `(account_id, channel)`, so no duplicates from re-running over old messages.
- **Performance**: no new IDB reads, no new RPC calls per tick. The added permission grant fires once per CDP session attach (small) and the matcher fallback is a `Vec::iter().find()` on the same target list we already enumerate.

## Related

- Closes #1016
- Follow-up PR(s)/TODOs:
  - file separate issues for huddle popup blank, parent CEF blank-after-huddle, dev:app restart_app regression, in-app logout sidebar UX
  - confirm whether the `(channel_id, day)` grouping intent in the issue body matters in practice once channels grow large; current code emits one doc per channel

## Summary by CodeRabbit

* **New Features**
  * Improved browser notification handling to grant permissions for webviews.

* **Bug Fixes**
  * Restored Slack memory ingestion and more reliable scanner attachment across navigation/pushState changes.
  * Fixed huddle popup and blanking issues in Slack webview for more stable audio/video interactions.

* **Documentation**
  * Added QA audit page detailing Slack parity tests and results.

### CODE REVIEW

**Comment by senamakel** [comment]
@oxoxDev pls resolve coderabbit issues

**Review by senamakel — CHANGES_REQUESTED** [review]
@oxoxDev pls resolve coderabbit issues

File: app/src-tauri/src/cdp/session.rs [hunk 1]
```diff
@@ -56,6 +56,35 @@ pub fn placeholder_url(account_id: &str) -> String {
...
+            }
+        }
+        idx += 1;
+    }
+    if slashes == 2 {
+        Some(url.to_string())
+    } else {
+        None
+    }
+}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply app/src-tauri/src/cdp/session.rs:86]: Fixed in 3e3ad9ef. `origin_of` now parses via `tauri::Url::parse()` and emits only `scheme://host[:port]` — handles paths/queries/fragments/userinfo/IPv6/case-normalization correctly and returns `None` for malformed input or non-http(s) schemes. Added 5 unit tests covering each edge case (cdp::session::tests::origin_of_*).

File: app/src-tauri/src/slack_scanner/mod.rs [hunk 2]
```diff
@@ -111,11 +111,24 @@ async fn scan_once(
...
     let page_target = targets
         .iter()
         .find(|t| {
             t.kind == "page" && t.url.starts_with(url_prefix) && t.url.ends_with(url_fragment)
         })
+        .or_else(|| {
+            targets
+                .iter()
+                .find(|t| t.kind == "page" && t.url.starts_with(url_prefix))
+        })
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply app/src-tauri/src/slack_scanner/mod.rs]: Fixed in 9c06be9c. Each scanner loop (IDB and DOM) now keeps a `pinned_target_id: Option<String>` that captures the CDP `targetId` on the first tick where the strict `#openhuman-account-<id>` fragment is still present. Subsequent ticks resolve in this order: pinned id -> strict fragment -> relaxed fallback. The relaxed fallback is intentionally never written back into the pin, so a misroute can't become permanent. The DOM path also tightens the relaxed fallback to require `/client` in the URL so we don't bind to a login redirect or marketing page. Applied at both call sites (`scan_once` and `dom_scan_once`).

File: docs/qa/SLACK-PARITY.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+- ⚠️ **partial** — works but with limitation; needs follow-up
+- ❌ **fail** — broken; child issue filed
+- 🔍 **needs investigation** — non-deterministic behavior; revisit
+- ⏭️ **skipped** — could not test (env / dependency missing)
+
+## Acceptance criteria audit
+
+| # | Criterion | Verdict | Evidence | Notes / child issue |
+|---|-----------|---------|----------|---------------------|
+| 1 | **Auth** — Google SSO, email/password, SAML SSO; session persists across app restarts | ✅ pass (login) / 🔍 (restart) | Vezures workspace loaded, signed-in as Nikhil Bajaj. Restart-persistence untestable in dev:app due to dev-mode restart-loop hack (separate child issue) | Login flow works through in-app webview. SAML not tested (no SAML org). Restart re-auth needs packaged .app to verify. |
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply docs/qa/SLACK-PARITY.md]: Fixed in 6a6c67cb. Row 1 now reads "needs to be validated in a packaged \`.app\`".

File: docs/qa/SLACK-PARITY.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
...7de6...","limit":20}` returns `{memories:[]}`. Log shows: `[webview-accounts] slack ScannerRegistry not in app state` immediately after Slack account opens. Zero `memory_doc_ingest` events ever fired. **Root cause**: `app/src-tauri/src/lib.rs:998` does `manage(std::sync::Arc::new(slack_scanner::ScannerRegistry::new()))` — but `slack_scanner::ScannerRegistry::new()` already returns `Arc<Self>` (`mod.rs:744-746`). Result: managed state is `Arc<Arc<ScannerRegistry>>`. Lookup at `webview_accounts/mod.rs:1751` is `try_state::<Arc<ScannerRegistry>>()` → returns None → scanner never spawns. WhatsApp/Discord/Telegram lines (997/999/1000) correctly use the bare `ScannerRegistry::new()` (no extra Arc wrap). Slack alone is double-wrapped. | **ONE-LINE FIX**. Change `lib.rs:998` from `.manage(std::sync::Arc::new(slack_scanner::ScannerRegistry::new()))` to `.manage(slack_scanner::ScannerRegistry::new())`. Post-fix: memory grouping per-day (audit dossier gap) becomes verifiable. Pre-audit "grouping by channel_id only" claim was wrong — `mod.rs:340` comment says per-(channel, day); will verify after fix. Reactions/threads extraction gaps still apply but blocked by getting scanner running first. |
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply docs/qa/SLACK-PARITY.md]: Fixed in 6a6c67cb. Row 7 criterion header dropped the stale `(channel_id, day)` claim and the closing note now reflects what `emit_and_persist` actually does — group by `channel_id` only, transcript carries each message's date inline. Per-day grouping stays as a deferred follow-up (also captured in the pre-audit gap list at the bottom of the doc). Also tightened the stale module-level comment in `slack_scanner/mod.rs` so source and doc agree.
